2.9BSD/usr/man/cat3/termcap.3

TERMCAP(3)          UNIX Programmer's Manual           TERMCAP(3)

NAME
     tgetent, tgetnum, tgetflag, tgetstr, tgoto, tputs - terminal
     independent operation routines

SYNOPSIS
     char PC;
     char *BC;
     char *UP;
     short ospeed;

     tgetent(bp, name)
     char *bp, *name;

     tgetnum(id)
     char *id;

     tgetflag(id)
     char *id;

     char *
     tgetstr(id, area)
     char *id, **area;

     char *
     tgoto(cm, destcol, destline)
     char *cm;

     tputs(cp, affcnt, outc)
     register char *cp;
     int affcnt;
     int (*outc)();

DESCRIPTION
     These functions extract and use capabilities from the termi-
     nal capability data base _t_e_r_m_c_a_p(5).  These are low level
     routines; see _c_u_r_s_e_s(3) for a higher level package.

     _T_g_e_t_e_n_t extracts the entry for terminal _n_a_m_e into the buffer
     at _b_p. _B_p should be a character buffer of size 1024 and must
     be retained through all subsequent calls to _t_g_e_t_n_u_m, _t_g_e_t_-
     _f_l_a_g, and _t_g_e_t_s_t_r. _T_g_e_t_e_n_t returns -1 if it cannot open the
     _t_e_r_m_c_a_p file, 0 if the terminal name given does not have an
     entry, and 1 if all goes well.  It will look in the environ-
     ment for a TERMCAP variable.  If found, and the value does
     not begin with a slash, and the terminal type name is the
     same as the environment string TERM, the TERMCAP string is
     used instead of reading the termcap file.  If it does begin
     with a slash, the string is used as a path name rather than
     /_e_t_c/_t_e_r_m_c_a_p. This can speed up entry into programs that
     call _t_g_e_t_e_n_t, as well as to help debug new terminal descrip-
     tions or to make one for your terminal if you can't write
     the file /_e_t_c/_t_e_r_m_c_a_p.

Printed 4/29/83                                                 1

TERMCAP(3)          UNIX Programmer's Manual           TERMCAP(3)

     _T_g_e_t_n_u_m gets the numeric value of capability _i_d, returning
     -1 if is not given for the terminal.  _T_g_e_t_f_l_a_g returns 1 if
     the specified capability is present in the terminal's entry,
     0 if it is not.  _T_g_e_t_s_t_r gets the string value of capability
     _i_d, placing it in the buffer at _a_r_e_a, advancing the _a_r_e_a
     pointer.  It decodes the abbreviations for this field
     described in _t_e_r_m_c_a_p(5), except for cursor addressing and
     padding information.

     _T_g_o_t_o returns a cursor addressing string decoded from _c_m to
     go to column _d_e_s_t_c_o_l in line _d_e_s_t_l_i_n_e. It uses the external
     variables UP (from the up capability) and BC (if bc is given
     rather than bs) if necessary to avoid placing \n, ^D or ^@
     in the returned string.  (Programs which call tgoto should
     be sure to turn off the XTABS bit(s), since tgoto may now
     output a tab.  Note that programs using termcap should in
     general turn off XTABS anyway since some terminals use con-
     trol I for other functions, such as nondestructive space.)
     If a % sequence is given which is not understood, then _t_g_o_t_o
     returns OOPS.

     _T_p_u_t_s decodes the leading padding information of the string
     _c_p; _a_f_f_c_n_t gives the number of lines affected by the opera-
     tion, or 1 if this is not applicable, _o_u_t_c is a routine
     which is called with each character in turn.  The external
     variable _o_s_p_e_e_d should contain the output speed of the ter-
     minal as encoded by _s_t_t_y (_2). The external variable PC
     should contain a pad character to be used (from the pc capa-
     bility) if a null (^@) is inappropriate.

FILES
     /usr/lib/libtermcap.a   -ltermcap library
     /etc/termcap    data base

SEE ALSO
     ex(1), curses(3), termcap(5)

AUTHOR
     William Joy

BUGS

Printed 4/29/83                                                 2



OpenBSD-4.6/gnu/usr.bin/perl/lib/termcap.pl
;# $RCSfile: termcap.pl,v $$Revision: 1.8 $$Date: 2008/09/29 17:36:10 $
#
# This library is no longer being maintained, and is included for backward
# compatibility with Perl 4 programs which may require it.
#
# In particular, this should not be used as an example of modern Perl
# programming techniques.
#
# Suggested alternative: Term::Cap
#
;#
;# Usage:
;#	require 'ioctl.pl';
;#	ioctl(TTY,$TIOCGETP,$foo);
;#	($ispeed,$ospeed) = unpack('cc',$foo);
;#	require 'termcap.pl';
;#	&Tgetent('vt100');	# sets $TC{'cm'}, etc.
;#	&Tputs(&Tgoto($TC{'cm'},$col,$row), 0, 'FILEHANDLE');
;#	&Tputs($TC{'dl'},$affcnt,'FILEHANDLE');
;#
sub Tgetent {
    local($TERM) = @_;
    local($TERMCAP,$_,$entry,$loop,$field);

    # warn "Tgetent: no ospeed set" unless $ospeed;
    foreach $key (keys %TC) {
	delete $TC{$key};
    }
    $TERM = $ENV{'TERM'} unless $TERM;
    $TERM =~ s/(\W)/\\$1/g;
    $TERMCAP = $ENV{'TERMCAP'};
    $TERMCAP = '/etc/termcap' unless $TERMCAP;
    if ($TERMCAP !~ m:^/:) {
	if ($TERMCAP !~ /(^|\|)$TERM[:\|]/) {
	    $TERMCAP = '/etc/termcap';
	}
    }
    if ($TERMCAP =~ m:^/:) {
	$entry = '';
	do {
	    $loop = "
	    open(TERMCAP,'<$TERMCAP') || die \"Can't open $TERMCAP\";
	    while (<TERMCAP>) {
		next if /^#/;
		next if /^\t/;
		if (/(^|\\|)${TERM}[:\\|]/) {
		    chop;
		    while (chop eq '\\\\') {
			\$_ .= <TERMCAP>;
			chop;
		    }
		    \$_ .= ':';
		    last;
		}
	    }
	    close TERMCAP;
	    \$entry .= \$_;
	    ";
	    eval $loop;
	} while s/:tc=([^:]+):/:/ && ($TERM = $1);
	$TERMCAP = $entry;
    }

    foreach $field (split(/:[\s:\\]*/,$TERMCAP)) {
	if ($field =~ /^\w\w$/) {
	    $TC{$field} = 1;
	}
	elsif ($field =~ /^(\w\w)#(.*)/) {
	    $TC{$1} = $2 if $TC{$1} eq '';
	}
	elsif ($field =~ /^(\w\w)=(.*)/) {
	    $entry = $1;
	    $_ = $2;
	    s/\\E/\033/g;
	    s/\\(200)/pack('c',0)/eg;			# NUL character
	    s/\\(0\d\d)/pack('c',oct($1))/eg;	# octal
	    s/\\(0x[0-9A-Fa-f][0-9A-Fa-f])/pack('c',hex($1))/eg;	# hex
	    s/\\(\d\d\d)/pack('c',$1 & 0177)/eg;
	    s/\\n/\n/g;
	    s/\\r/\r/g;
	    s/\\t/\t/g;
	    s/\\b/\b/g;
	    s/\\f/\f/g;
	    s/\\\^/\377/g;
	    s/\^\?/\177/g;
	    s/\^(.)/pack('c',ord($1) & 31)/eg;
	    s/\\(.)/$1/g;
	    s/\377/^/g;
	    $TC{$entry} = $_ if $TC{$entry} eq '';
	}
    }
    $TC{'pc'} = "\0" if $TC{'pc'} eq '';
    $TC{'bc'} = "\b" if $TC{'bc'} eq '';
}

@Tputs = (0,200,133.3,90.9,74.3,66.7,50,33.3,16.7,8.3,5.5,4.1,2,1,.5,.2);

sub Tputs {
    local($string,$affcnt,$FH) = @_;
    local($ms);
    if ($string =~ /(^[\d.]+)(\*?)(.*)$/) {
	$ms = $1;
	$ms *= $affcnt if $2;
	$string = $3;
	$decr = $Tputs[$ospeed];
	if ($decr > .1) {
	    $ms += $decr / 2;
	    $string .= $TC{'pc'} x ($ms / $decr);
	}
    }
    print $FH $string if $FH;
    $string;
}

sub Tgoto {
    local($string) = shift(@_);
    local($result) = '';
    local($after) = '';
    local($code,$tmp) = @_;
    local(@tmp);
    @tmp = ($tmp,$code);
    local($online) = 0;
    while ($string =~ /^([^%]*)%(.)(.*)/) {
	$result .= $1;
	$code = $2;
	$string = $3;
	if ($code eq 'd') {
	    $result .= sprintf("%d",shift(@tmp));
	}
	elsif ($code eq '.') {
	    $tmp = shift(@tmp);
	    if ($tmp == 0 || $tmp == 4 || $tmp == 10) {
		if ($online) {
		    ++$tmp, $after .= $TC{'up'} if $TC{'up'};
		}
		else {
		    ++$tmp, $after .= $TC{'bc'};
		}
	    }
	    $result .= sprintf("%c",$tmp);
	    $online = !$online;
	}
	elsif ($code eq '+') {
	    $result .= sprintf("%c",shift(@tmp)+ord($string));
	    $string = substr($string,1,99);
	    $online = !$online;
	}
	elsif ($code eq 'r') {
	    ($code,$tmp) = @tmp;
	    @tmp = ($tmp,$code);
	    $online = !$online;
	}
	elsif ($code eq '>') {
	    ($code,$tmp,$string) = unpack("CCa99",$string);
	    if ($tmp[$[] > $code) {
		$tmp[$[] += $tmp;
	    }
	}
	elsif ($code eq '2') {
	    $result .= sprintf("%02d",shift(@tmp));
	    $online = !$online;
	}
	elsif ($code eq '3') {
	    $result .= sprintf("%03d",shift(@tmp));
	    $online = !$online;
	}
	elsif ($code eq 'i') {
	    ($code,$tmp) = @tmp;
	    @tmp = ($code+1,$tmp+1);
	}
	else {
	    return "OOPS";
	}
    }
    $result . $string . $after;
}

1;
